### PR TITLE
add 'client_request_total' to statistics request count of clients in …

### PR DESCRIPTION
**Pull request by kkkkun**
#### What type of PR is this?
/kind flake

#### What this PR does / why we need it:
add new metric client_request_count.

When we use more operators, apiserver will be requested by operator. We hope that all clients use listwatch to request apiserver. But some operators do not so, they still get and list by round robin. It will increase pressure to apiserver.
In our kubernetes's environment, some operators list or get apiserver frequently, which lead to apiserver's heap increase in one second, and then garbage collection starts to work. It's hard to find out operator which do not use listwatch.
Based on the above background, it is neccessary to add a new metric that record client's request count of list and get.
It will find out nonstandard clients easily.

### CODE REVIEW

**Comment by dgrisonnet** [comment]
You should already be able to identify which operators aren't using listwatchers by looking into the audit logs exposed by the apiserver: https://kubernetes.io/docs/tasks/debug-application-cluster/audit/

One of the reasons behind not having the client information in the `apiserver_requests_total` total metric is that this information wouldn't be useful from a backend perspective. Also, this metric is already very expensive so we can't just add new dimensions to it. As such, one usually goes through the audit logs to figure out the origin of the requests sent to the apiserver.

Also, generally,  I don't think it is right to have client-side metrics exposed by the backend. If deemed useful, this kind of metric should be exposed on the client-side rather than by the backend.

/hold

**Comment by kkkkun** [comment]
> You should already be able to identify which operators aren't using listwatchers by looking into the audit logs exposed by the apiserver: https://kubernetes.io/docs/tasks/debug-application-cluster/audit/
> 
> One of the reasons behind not having the client information in the `apiserver_requests_total` total metric is that this information wouldn't be useful from a backend perspective. Also, this metric is already very expensive so we can't just add new dimensions to it. As such, one usually goes through the audit logs to figure out the origin of the requests sent to the apiserver.
> 
> Also, generally, I don't think it is right to have client-side metrics exposed by the backend. If deemed useful, this kind of metric should be exposed on the client-side rather than by the backend.
> 
> /hold

Thank you for your reply !
If we set client with request's userAgent,  it is very expensive.  So we just set client's name which do not containe ip address, it will decrease count of metric. and it's  very useful to point out  client which often list or get resources. 
Otherwise,  If we look into the audit logs， it still need to be dealed with special tool. And it will produce a large number of logs.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go [hunk 1]
```diff
@@ -240,6 +240,16 @@ var (
 		[]string{"source", "status"},
 	)
 
+	//clientRequestCounter is a number of total requests with client name
+	clientRequestCounter = compbasemetrics.NewCounterVec(
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:244]: We already have a metric for request counter:
https://github.com/kubernetes/kubernetes/pull/106216/files#diff-c4f07ee44a4626c817dd95cf30e147a2125ea923804be5c3608fb4a845fc1c20R74

We used to have a client label there, but that was actually a vulnerability, as the client is a potentially unbounded dimensions.

So we don't want to add that.

@logicalhan - FYI

Comment (kkkkun) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:244]: That makes sense.  We will try to look out operator which do not use listwatch by other method.

Thanks!

**Comment by tkashem** [comment]
it's not a global solution, but operators in question can implement component-base metric so it has the client-go metrics available, just like scheduler, controller-manager and kubelet.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go (hunk 1 above)
Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:244]: Yeah we can't do this... client is unbounded and can be manipulated in the request; it is therefore an attack vector.

**Comment by logicalhan** [comment]
You should query logs for this type of information, imo.

**Comment by dgrisonnet** [comment]
I agree with @logicalhan, logs cover this kind of use case perfectly.

> Otherwise, If we look into the audit logs， it still need to be dealed with special tool. And it will produce a large number of logs.

The goal of having multiple observability signals is exactly to separate information based on their granularity. Metrics are constantly evaluated so you can't afford to expose a lot of information inside of them. You should stick to the bare minimum that would allow you to do alerting, create dashboards, investigate incidents and get some insights on what is running inside of your cluster.
If you want to get precise information about the queries in your cluster, such as their origin and type, logs are better suited since they allow you to store more data and are evaluated on an as-needed basis.
If you want to know exactly what is happening inside of your application at a certain time then traces are a better vector of information.

So yeah, if you want more granular information about your cluster, the best is to look into how to consume that new signal rather than trying to fit information in an unsuited signal that will end up costing you way more.

File: staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go (hunk 1 above)
Comment (kkkkun) [reply staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:244]: ok，thanks

**Comment by kkkkun** [comment]
> I agree with @logicalhan, logs cover this kind of use case perfectly.
> 
> > Otherwise, If we look into the audit logs， it still need to be dealed with special tool. And it will produce a large number of logs.
> 
> The goal of having multiple observability signals is exactly to separate information based on their granularity. Metrics are constantly evaluated so you can't afford to expose a lot of information inside of them. You should stick to the bare minimum that would allow you to do alerting, create dashboards, investigate incidents and get some insights on what is running inside of your cluster. If you want to get precise information about the queries in your cluster, such as their origin and type, logs are better suited since they allow you to store more data and are evaluated on an as-needed basis. If you want to know exactly what is happening inside of your application at a certain time then traces are a better vector of information.
> 
> So yeah, if you want more granular information about your cluster, the best is to look into how to consume that new signal rather than trying to fit information in an unsuited signal that will end up costing you way more.

ok， I will try to statistics by logs. 

This PR will be closed.

**Comment by kkkkun** [comment]
This PR will be closed.  
Thank all!

**Pull request closed without merging**